admin/src/main/scala/slamdata/engine/admin/table.scala

```
@@ -106,32 +110,36 @@ object Values {
   def mergePaths(as: List[Path], bs: List[Path]): List[Path] =
     (as ++ bs).distinct
   
-  def flatten(json: Json): ListMap[Path, String] = {
-    def loop(json: Json): String \/ List[(Path, String)] = {
-      def prepend(name: String, json: Json): List[(Path, String)] = 
-        loop(json) match {
+  def flatten(data: Data): ListMap[Path, Data] = {
+    def loop(data: Data): Data \/ List[(Path, Data)] = {
+      def prepend(name: String, data: Data): List[(Path, Data)] = 
+        loop(data) match {
           case -\/ (value) => (List(name) -> value) :: Nil
           case  \/-(map)   => map.map(t => (name :: t._1) -> t._2)
         }
-      json.fold(
-        -\/(""),
-        bool => -\/ (bool.toString),
-        num  => -\/ (json.toString),
-        str  => -\/ (str),
-        arr  =>  \/-(arr.zipWithIndex.flatMap { case (c, i) => prepend(i.toString, c) }),
-        obj  =>  obj.toList.flatMap { case (f, c) => prepend(f, c) } match {
-          case (List("$date"), value: String) :: Nil => -\/(value)
-          case (List("$oid"), value: String) :: Nil => -\/(value)
-          case map => \/-(map)
-        }
-      )
+      data match {
+        case Data.Arr(value) =>  \/-(value.zipWithIndex.flatMap { case (c, i) => prepend(i.toString, c) })
+        case Data.Obj(value) =>  \/-(value.toList.flatMap { case (f, c) => prepend(f, c) })
+        case _               => -\/ (data)
+      }
     }
     
-    loop(json) match {
+    loop(data) match {
       case -\/ (value) => ListMap(List("value") -> value)
       case  \/-(map)   => map.toListMap
     }
   }
+  
+  def renderSimple(data: Data)(implicit C: DataCodec) = {
+    def loop(json: Json): String = json.fold(
+        json.toString,
+        κ(json.toString),  // NB: prints ints without trailing zero (unlike Double's toString, apparently)
+        κ(json.toString),
+        identity,  // No surrounding quotes on string values
+        arr => "unexpected array: " + arr,
```

This rendering and the one beneath it is invalid. We should propagate the error higher or remove the potential for error by modifying the type signature of `loop` to only accept things for which it provably has a simple string rendering.